### PR TITLE
Refactor TraceAnalyzer to use `TraceRecord::Handler` to avoid casting.

### PR DESCRIPTION
**Pull request by autopear**
`TraceAnalyzer` privately inherits `TraceRecord::Handler` and `WriteBatch::Handler`.

`trace_analyzer_test` can pass with this change.

### CODE REVIEW

File: tools/trace_analyzer_tool.cc [hunk 1]
```diff
@@ -469,96 +472,29 @@ Status TraceAnalyzer::StartProcessing() {
       break;
     }
 
+    // To do: why increment total_requests_ for trace end?
```
Comment (zhichao-cao) [inline tools/trace_analyzer_tool.cc:475]: Sometimes, the trace might be incomplete. We may not be able to the type::end.

File: tools/trace_analyzer_tool.h [hunk 2]
```diff
@@ -182,24 +183,64 @@ class TraceAnalyzer {
...
+                std::unique_ptr<TraceRecordResult>* result) override;
+  Status Handle(const GetQueryTraceRecord& record,
+                std::unique_ptr<TraceRecordResult>* result) override;
+  Status Handle(const IteratorSeekQueryTraceRecord& record,
+                std::unique_ptr<TraceRecordResult>* result) override;
+  Status Handle(const MultiGetQueryTraceRecord& record,
+                std::unique_ptr<TraceRecordResult>* result) override;
+
+  using WriteBatch::Handler::PutCF;
+  Status PutCF(uint32_t column_family_id, const Slice& key,
```
Comment (zhichao-cao) [inline tools/trace_analyzer_tool.h:200]: What does "CF" means here? CallFunction? or ColumnFamily?

Comment (autopear) [reply tools/trace_analyzer_tool.h:200]: It's ColumnFamily, from
https://github.com/facebook/rocksdb/blob/ff8953380f356768d5ea7bf26b2734d86484d228/include/rocksdb/write_batch.h#L182-L193

File: tools/trace_analyzer_tool.cc (hunk 1 above)
Comment (autopear) [reply tools/trace_analyzer_tool.cc:475]: Yes, my point is, you should not increment total_requests if you read a `TraceEnd`, so it should be 
```
    end_time_ = trace.ts;
    if (trace.type == kTraceEnd) {
      break;
    }
    total_requests_++;
```
Then `end_time_` is the timetamp of the last trace (or trace end), and `total_requests_` will be the actual number of requests.

File: tools/trace_analyzer_tool.h [hunk 3]
```diff
@@ -213,6 +254,7 @@ class TraceAnalyzer {
   uint64_t total_access_keys_;
   uint64_t total_gets_;
   uint64_t total_writes_;
+  // To do: why no counters for seeks and multigets?
```
Comment (zhichao-cao) [inline tools/trace_analyzer_tool.h:257]: Good catch! I forgot to add them........

File: trace_replay/trace_replay.cc [hunk 4]
```diff
@@ -126,222 +126,218 @@ bool TracerHelper::SetPayloadMap(uint64_t& payload_map,
...
+    case kTraceIteratorSeekForPrev: {
+      uint32_t cf_id = 0;
+      Slice iter_key;
+      Slice lower_bound;
+      Slice upper_bound;
+
+      if (trace_file_version < 2) {
+        DecodeCFAndKey(trace->payload, &cf_id, &iter_key);
+      } else {
+        Slice buf(trace->payload);
```
Comment (zhichao-cao) [inline trace_replay/trace_replay.cc:225]: It is not that elegant to put the payload decoding in the trace analyzer.......

File: tools/trace_analyzer_tool.cc (hunk 1 above)
Comment (zhichao-cao) [reply tools/trace_analyzer_tool.cc:475]: > Yes, my point is, you should not increment total_requests if you read a `TraceEnd`, so it should be
> 
> ```
>     end_time_ = trace.ts;
>     if (trace.type == kTraceEnd) {
>       break;
>     }
>     total_requests_++;
> ```
> 
> Then `end_time_` is the timetamp of the last trace (or trace end), and `total_requests_` will be the actual number of requests.

Sure, just follow what you proposed above.

File: trace_replay/trace_replay.cc (hunk 4 above)
Comment (autopear) [reply trace_replay/trace_replay.cc:225]: This is in TracerHelper, not in TraceAnalyzer. I merged DecodeWriteTraceRecord / DecodeGetTraceRecord, ... into a single `DecodeTraceRecord`, then you don't need to call each individual decoding functions in `TraceAnalyzer` and `ReplayerImpl`.

File: tools/trace_analyzer_tool.cc (hunk 1 above)
Comment (autopear) [reply tools/trace_analyzer_tool.cc]: Done

File: tools/trace_analyzer_tool.h (hunk 3 above)
Comment (autopear) [reply tools/trace_analyzer_tool.h]: Added

File: trace_replay/trace_replay.cc (hunk 4 above)
Comment (zhichao-cao) [reply trace_replay/trace_replay.cc:225]: Sorry, I did not notice that....

**Review by zhichao-cao — APPROVED** [review]
LGTM, thank you so much for the refactor work, excellent job!

**Pull request closed without merging**